[arc-opts] Teach IsAddressWrittenToDefUseAnalysis how to track "well behaved" writes but don't use it.

gottesmm opened the pull request:
In a future commit, I am going to build on this to promote load [copy] ->
load_borrow from inout arguments where none of the writes overlap the load
[copy]'s result's lifetime. By committing this separately, I am using the
current pass's logic to validate the change.

gottesmm committed:
[arc-opts] Teach IsAddressWrittenToDefUseAnalysis how to track "well behaved" writes but don't do use it.

In a future commit, I am going to build on this to promote load [copy] ->
load_borrow from inout arguments where none of the writes overlap the load
[copy]'s result's lifetime. By committing this separately, I am using the
current pass's logic to validate the change.

gottesmm commented:
@swift-ci test

atrick commented on lib/SILOptimizer/Transforms/SemanticARCOpts.cpp:
```diff
@@ -482,30 +483,37 @@ namespace {
-private:
-  bool isWrittenToHelper(SILValue value);
+struct IsAddressWrittenToDefUseAnalysis
+    : public SmallMultiMapCache<IsAddressWrittenToDefUseAnalysis, SILValue,
```
Please explain how the Analysis is supposed to be viewed as a cache. What does the `SmallMultiMapCache` interface mean when it's inherited by this analysis. Or just don't publicly inherit a generic ADT and give the analysis its own interface.

atrick commented on lib/SILOptimizer/Transforms/SemanticARCOpts.cpp:
```diff
@@ -1659,7 +1677,8 @@ class StorageGuaranteesLoadVisitor
     // never actually written to, then we convert to a load_borrow.
-    return answer(ARCOpt.isAddressWrittenToDefUseAnalysis(access));
+    auto result = ARCOpt.isAddressWrittenToDefUseAnalysis.get(access);
+    return answer(!result.hasValue() || result.getValue().size());
```
I can't make sense of this code since it's written in terms of the multi-map interface. What does it mean for the analysis to have or not have a value, and how is that different from it being empty or having multiple elements?

atrick reviewed (commented):
Fairly enigmatic code

atrick commented:
Backing up, my main issue here is "multi-map cache", which was in a different PR. A cache cannot be a multi-map by definition. Either the analysis has a valid result or it does not. If the result is comprised of several values, that is distinct from whether the result is valid or complete.

gottesmm committed:
[multimapcache] Change multi-map cache to use a std::function instead of CRTP.

This makes it so one uses a passed in std::function, instead of an impl class to
map a key to a list of values to be cached.

gottesmm commented:
@atrick I fixed the cache as per our discussion.

gottesmm commented:
@swift-ci test

gottesmm commented:
Windows failed in a c++ interior test. Not this commit.
